libbpf-tools/biostacks.c

```diff
@@ -172,52 +172,56 @@ int main(int argc, char **argv)
 
 	obj->rodata->targ_ms = env.milliseconds;
 
-	if (fentry_can_attach("blk_account_io_start", NULL)) {
-		bpf_program__set_attach_target(obj->progs.blk_account_io_start, 0,
-					       "blk_account_io_start");
-		bpf_program__set_attach_target(obj->progs.blk_account_io_done, 0,
-					       "blk_account_io_done");
+	if (kprobe_exists("blk_account_io_start")) {
```

Comment:
The userspace part looks way to complicated, could you please simplify that ?

Let's first check if fentry is supported:
* if yes, disable all kprobe progs, and do auto-attach
* if no, disable all fentry progs, and do auto-attach
You may need to disable some progs if symbols not exist

Reply:
Same for #4549 and #4551.

Reply:
donw

Reply:
done

Reply:
Hi @chenhengqi
I did some simplification follow you suggestion, do you have some more advice?
And there is a failure in https://github.com/iovisor/bcc/pull/455, but I cann't see the cause of that request?